### PR TITLE
Introduce HerdDB Database as cache over ZK metadata

### PR DESCRIPTION
**Pull request by eolivelli**
See #32 as master ticket.

Contents of this first patch:
- Add HerdDB JDCB drivers (this is enough to start the server in 'embedded' mode
- Add EclipseLink JPA (Java Persistence API) implementation and setup a minimal JPA PersistenceUnit
- With JPA the database schema is generated automatically, no need to write DDL, supports automatically every supported DB by EclipseLink
- Add test suite for JPA part
- Store LedgerMetadata on DB in order to save resources on ZK

Known issues to be resolved as follow up:
- database is kept in memory: we have to add configuration for DB, at least the JDBC URL and username/password, with JDBC URL you can control the JDBC driver and decide to start the server in the same process or connect to an external one
- the 'Cache' page is really ugly and we should have a button "Refresh metadata" with at least a Spinner icon (metadata refresh can take minutes in a production cluster)
- there is only one "LEDGER" table that holds data, we have to at least add a "LEDGER_METADATA" table in order to perform fast searchs, like application=pulsar, topic=path/to/topic

### CODE REVIEW

**Comment by eolivelli** [comment]
This is only a draft

File: checkstyle.xml [hunk 1]
```diff
@@ -224,22 +224,7 @@ page at http://checkstyle.sourceforge.net/config.html -->
         LENGTH and CODING CHECKS
 
         -->
-
-        <module name="LineLength">
```
Comment (eolivelli) [inline checkstyle.xml]: We dropped this rule from the other projects

File: src/main/java/org/bookkeepervisualmanager/bookkeeper/BookkeeperManager.java [hunk 2]
```diff
@@ -79,11 +91,37 @@ public BookkeeperManager(ConfigurationStore configStore) throws BookkeeperExcept
...
+        LOG.info("Refreshing Metadata Cache");
+        try {
+            Iterable<Long> ledgersIds = bkAdmin.listLedgers();
+            for (long ledgerId : ledgersIds) {
+                LedgerMetadata ledgerMetadata = readLedgerMetadata(ledgerId);
+                Ledger ledger = new Ledger(ledgerId,
+                        ledgerMetadata.getLength(),
+                        new java.sql.Timestamp(ledgerMetadata.getCtime()),
+                        new java.sql.Timestamp(System.currentTimeMillis()),
+                        Base64.getEncoder().encodeToString(serDe.serialize(ledgerMetadata)));
```
Comment (eolivelli) [inline src/main/java/org/bookkeepervisualmanager/bookkeeper/BookkeeperManager.java]: ZK metadata is serialized to a byte[] by the native BK API (serDe), then we are storing it in base64 enconding on DB

File: src/main/java/org/bookkeepervisualmanager/bookkeeper/BookkeeperManager.java [hunk 3]
```diff
@@ -117,16 +152,46 @@ public LedgerManager getLedgerManager() {
 
     public List<Long> getLedgersForBookie(String bookieId) throws BookkeeperException {
```
Comment (eolivelli) [inline src/main/java/org/bookkeepervisualmanager/bookkeeper/BookkeeperManager.java]: now we are using the DB to compute this list, no more need to scan the entire ZK tree

File: src/main/java/org/bookkeepervisualmanager/bookkeeper/BookkeeperManager.java [hunk 4]
```diff
@@ -149,16 +214,12 @@ public LedgerMetadata getLedgerMetadata(long ledgerId) throws BookkeeperExceptio
     }
 
     public List<Long> getAllLedgers() throws BookkeeperException {
-        try {
-            List<Long> resultLedgers = new ArrayList<>();
-
-            Iterable<Long> ledgersIds = bkAdmin.listLedgers();
-            ledgersIds.forEach(resultLedgers::add);
+        return metadataCache
```
Comment (eolivelli) [inline src/main/java/org/bookkeepervisualmanager/bookkeeper/BookkeeperManager.java]: we could return more data to the rest API, in order to show more info on the page that lists the ledgers

**Comment by mino181295** [comment]
I'm not really into EclipseLink JPA but this patch looks okay to me. You can merge and proceed

**Comment by eolivelli** [comment]
Thanks

**Pull request merged**